File: Brute Force/Heapsort/code_edu.java
import org.algorithm_visualizer.*;
import java.util.Arrays;

public class Main {
    //visualizer{
	private static ChartTracer chart = new ChartTracer();
	private static LogTracer logger = new LogTracer("Console");
	private static Array1DTracer tracer = new Array1DTracer("Array1DTracer");
	//}
	private static int sizeOfArray = /*배열 크기 입력*/
    private static int minValue = /*배열 최소값 입력*/
    private static int maxValue = /*배열 최대값 입력*/
    
	//array[] :sizeOfArray개의 minValue ~ maxValue 사이의 난수 값을 갖는 배열
    private static Integer [] array = (Integer[]) new Randomize.Array1D(sizeOfArray, new Randomize.Integer(minValue, maxValue)).create();
	Main()
	{
	    //visualizer{
		Layout.setRoot(new VerticalLayout(new Commander[]{chart, tracer, logger}));	
		tracer.set(array);
		tracer.chart(chart);
        //}
		heapSort(array, array.length);
		
		for(int i =0; i < array.length; i++)
			logger.print(array[i] + " ");
	}

	void heapSort(Integer[] array, int size) 
	{
		int i, j, temp;

		for (i = (int) (Math.ceil(size / 2) - 1); i >= 0; i--) 
		{
			heapify(array, size, i);
		}

		for (j = size - 1; j >= 0; j--) 
		{
			temp = array[0];
			array[0] = array[j];
			array[j] = temp;
            
            //visualizer{
			tracer.patch(0, array[0]);
			tracer.patch(j, array[j]);
			logger.println("Swapping elements : " + array[0] + " & " +  array[j]);
			Tracer.delay();
			tracer.depatch(0);
			tracer.depatch(j);
			tracer.select(j);
			Tracer.delay();
            //}
			heapify(array, j, 0);
            //visualizer{
			tracer.deselect(j);
			//}
		}
	}

	void heapify(Integer[] array, int size, int root) 
	{
		int largest = root;
		int left = /*알맞은 값*/
		int right = /*알맞은 값*/
		int temp;

		if (left < size && /*알맞은 조건*/)
		{
			largest = left;
		}

		if (right < size && /*알맞은 조건*/) 
		{
			largest = right;
		}

		if (largest != root) 
		{
		    //swap ( array[root], array[largest] )
			temp = /*?*/
			/*?*/ = /*?*/
			/*?*/ = /*?*/
			
            //visualizer{
			tracer.patch(root, array[root]);
			tracer.patch(largest, array[largest]);
			logger.println("Swapping elements : " + array[root] + " & " + array[largest]);
			Tracer.delay();
			tracer.depatch(root);
			tracer.depatch(largest);
            //}
			heapify(array, size, largest);
		}
	}

	public static void main(String[] args) {
		new Main();
	}
}

File: Brute Force/Heapsort/code_edu.cpp
#include "algorithm-visualizer.h"
#include <iostream>

const int				ARRAYSIZE = /*(배열 길이 입력)*/

//visualizer{
ChartTracer				chart("ChartTracer");
Array1DTracer			tracer("Array1DTracer");
LogTracer				logger("LogTracer");
Randomize::Array1D<int> array1D(ARRAYSIZE);
int						D[ARRAYSIZE];
int						count = 0; //무한 재귀호출 및 루프 방지

void InitHeapSort()
{
	Layout::setRoot(VerticalLayout({ chart,tracer,logger }));
	array1D.fill(D);
	tracer.set(D);
	tracer.chart(chart);
	Tracer::delay();
}
// }

void Heapify(int _array[ARRAYSIZE], int size, int root)
{
	//정렬할 배열구조를 힙으로 만드는 함수

	//visualizer{
	if (count++ > ARRAYSIZE*ARRAYSIZE*2) return;	//무한루프 방지
	//}
	int largest = root;
	int left = 2 * root + 1;
	int right = /*(알맞은 값 입력)*/
	int temp;

	if (left < size && _array[left] > _array[largest]) {
		largest = /*(알맞은 값 입력)*/
	}

	if (right <size && _array[right] > _array[largest]) {
		largest = /*(알맞은 값 입력)*/
	}

	if (largest != root) {
		temp = _array[root];
		_array[root] = _array[largest];
		_array[largest] = temp;

		//visualizer{
		tracer.patch(root, _array[root]);
		tracer.patch(largest, _array[largest]);

		logger.println("Swapping elements : " + std::to_string(_array[root]) + " and " + std::to_string(_array[largest]));
		Tracer::delay();
		tracer.depatch(root);
		tracer.depatch(largest);
		//}
		Heapify(_array, size, /*(알맞은 값 입력)*/);
	}
}

void HeapSort(int _array[ARRAYSIZE], int size)
{
	int i, j, temp;

	for (i = size / 2; i >= 0; i--)
		Heapify(_array, size, i);

	for (j = size - 1; j >= 0; j--) {
		temp = _array[0];
		_array[0] = _array[j];
		_array[j] = temp;

		//visualizer{
		tracer.patch(0, _array[0]);
		tracer.patch(j, _array[j]);
		logger.println("Swapping element : " + std::to_string(_array[0]) + " and " + std::to_string(_array[j]));
		Tracer::delay();
		tracer.depatch(0);
		tracer.depatch(j);
		tracer.select(j);
		Tracer::delay();
		//}
		Heapify(_array, /*(알맞은 값 입력)*/, 0);

		//visualizer{
		tracer.deselect(j);
		//}
	}
}

int main()
{
	InitHeapSort();
	//visualizer{
	logger.print("original array = [");
	for (int i = 0; i < ARRAYSIZE-1; i++)
	{
		logger.print(std::to_string(D[i]) + " , ");
	}
	logger.println(std::to_string(D[ARRAYSIZE-1]) + " ] ");
	//}
	HeapSort(D, sizeof(D) / sizeof(int));

	//visualizer{
	logger.print("Final array = [");
	for (int i = 0; i < ARRAYSIZE-1; i++)
	{
		logger.print(std::to_string(D[i]) + " , ");
	}
	logger.println(std::to_string(D[ARRAYSIZE-1]) + " ] ");
	//}
}